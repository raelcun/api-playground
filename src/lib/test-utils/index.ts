import * as HttpStatus from 'http-status-codes'
import { Context } from 'koa'

import { Logger } from '@lib/logger'
import { install, InstalledClock } from '@sinonjs/fake-timers'

export const createMockLogger = (): Logger => ({
  info: jest.fn(),
  fatal: jest.fn(),
  trace: jest.fn(),
  error: jest.fn(),
  debug: jest.fn(),
  warn: jest.fn(),
  child: jest.fn(),
})

export const testUnauthorized = (mockContext: Context, mockNext: jest.Mock<Promise<void>, []>) => {
  test('should set status to 401', () => {
    expect(mockContext.status).toEqual(HttpStatus.UNAUTHORIZED)
  })

  test('should not invoke next', () => {
    expect(mockNext).not.toHaveBeenCalled()
  })
}

export const testAuthorized = (mockContext: Context, mockNext: jest.Mock<Promise<void>, []>) => {
  test('should not set status to 401', () => {
    expect(mockContext.status).not.toEqual(HttpStatus.UNAUTHORIZED)
  })

  test('should invoke next once', () => {
    expect(mockNext).toHaveBeenCalledTimes(1)
  })
}

/* istanbul ignore next */
export const installStaticClock = () => {
  let clock: InstalledClock

  beforeAll(() => {
    clock = install({ now: 946684800000 })
  })

  afterAll(() => {
    clock.uninstall()
  })

  return () => clock
}
