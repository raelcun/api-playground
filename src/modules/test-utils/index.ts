import { sign } from 'jsonwebtoken'

import { Roles } from '@lib/rbac'
import { Token } from '@modules/api-core'
import { getConfig } from '@modules/config'

const JWT_SECRET = getConfig().server.jwtSecret

export const createExpiredTokenHeader = () =>
  `Bearer ${sign(<Token>{ userId: 'foo', role: Roles.User }, JWT_SECRET, {
    expiresIn: '-10s',
  })}`

export const createInvalidTokenSchemaHeader = () => `Bearer ${sign({ userId: 'foo' }, JWT_SECRET)}`

export const createValidUserTokenHeader = () =>
  `Bearer ${sign(<Token>{ userId: 'foo', role: Roles.User }, JWT_SECRET)}`

export const createValidAdminTokenHeader = () =>
  `Bearer ${sign(<Token>{ userId: 'foo', role: Roles.Admin }, JWT_SECRET)}`
