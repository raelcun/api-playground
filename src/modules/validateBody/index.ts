export { validateBody, withValidatedBody } from './validateBody'
