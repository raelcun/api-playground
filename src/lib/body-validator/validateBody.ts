import { either as E } from 'fp-ts'
import { pipe } from 'fp-ts/lib/pipeable'
import t from 'io-ts'

import { Err } from '@lib/error'
import { logErrorsE, LoggerProvider } from '@lib/logger'
import { decode, mapErrorCode } from '@lib/utils'

export const validateBody = (getLogger: LoggerProvider) => <T>(type: t.Type<T, unknown>) => (
  body: unknown,
): E.Either<Err<'BODY_VALIDATION_ERROR'>, T> =>
  pipe(decode(type, body), logErrorsE(getLogger()), mapErrorCode('BODY_VALIDATION_ERROR' as const))
