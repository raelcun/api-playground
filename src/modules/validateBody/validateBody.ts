import { either as E } from 'fp-ts'
import { pipe } from 'fp-ts/lib/pipeable'
import HttpStatus from 'http-status-codes'
import t from 'io-ts'
import { Middleware as KoaMiddleware } from 'koa'

import { createErrorResponse } from '@modules/api-core/response'
import { Err } from '@modules/error/types'
import { getSystemLogger } from '@modules/logger'
import { LoggerFactory } from '@modules/logger/types'
import { KoaContext, Middleware } from '@root/types'
import { decode, logErrors, mapErrorCode } from '@utils'

export const validateBodyInner = (createLogger: LoggerFactory) => <T>(type: t.Type<T, unknown>) => (
  body: unknown,
): E.Either<Err<'BODY_VALIDATION_ERROR'>, T> =>
  pipe(
    decode(type, body),
    mapErrorCode('BODY_VALIDATION_ERROR' as const),
    logErrors(createLogger()),
  )

export const validateBody = validateBodyInner(getSystemLogger)

export const withValidatedBodyInner = (createLogger: LoggerFactory) => <T>(
  type: t.Type<T, unknown>,
) => (middleware: Middleware<T>): KoaMiddleware => async (ctx, next) => {
  const result = validateBodyInner(createLogger)(type)(ctx.request.body)

  if (E.isLeft(result)) {
    ctx.status = HttpStatus.BAD_REQUEST
    ctx.body = createErrorResponse(result.left)
    return
  }

  await middleware(ctx as KoaContext<T>, next)
}

export const withValidatedBody = withValidatedBodyInner(getSystemLogger)
