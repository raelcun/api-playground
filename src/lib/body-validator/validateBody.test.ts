import * as t from 'io-ts'

import { createMockLogger } from '@lib/test-utils'

import { validateBody } from './validateBody'

describe('validateBody', () => {
  describe('when validation is successful', () => {
    test('should return input', () => {
      const result = validateBody(createMockLogger)(t.string)('foo')

      expect(result).toMatchSnapshot()
    })
  })

  describe('when validation is unsuccessful', () => {
    test('should return error', () => {
      const result = validateBody(createMockLogger)(t.string)(5)

      expect(result).toMatchSnapshot()
    })

    test('should log errors', () => {
      const logger = createMockLogger()
      const logSpy = jest.spyOn(logger, 'log')
      validateBody(() => logger)(t.string)(5)

      expect(logSpy.mock.calls).toMatchSnapshot()
    })
  })
})
