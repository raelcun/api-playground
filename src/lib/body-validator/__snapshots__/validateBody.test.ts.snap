// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateBody when validation is successful should return input 1`] = `
Object {
  "_tag": "Right",
  "right": "foo",
}
`;

exports[`validateBody when validation is unsuccessful should log errors 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "code": "BODY_VALIDATION_ERROR",
        "message": "Expecting string but instead got: 5",
      },
    },
  ],
]
`;

exports[`validateBody when validation is unsuccessful should return error 1`] = `
Object {
  "_tag": "Left",
  "left": Object {
    "code": "BODY_VALIDATION_ERROR",
    "message": "Expecting string but instead got: 5",
  },
}
`;
